src/gui/widgets/LcdFloatSpinBox.cpp

```diff
@@ -0,0 +1,255 @@
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "LcdFloatSpinBox.h"
+
+#include <cmath>
+#include <QApplication>
+#include <QHBoxLayout>
+#include <QInputDialog>
+#include <QLabel>
+#include <QMouseEvent>
+#include <QPainter>
+#include <QPixmap>
+#include <QStyleOptionFrameV2>
+#include <QVBoxLayout>
+
+#include "CaptionMenu.h"
+#include "embed.h"
+#include "GuiApplication.h"
+#include "gui_templates.h"
+#include "MainWindow.h"
+
+
+LcdFloatSpinBox::LcdFloatSpinBox(int numWhole, int numFrac, QWidget* parent, const QString& name) :
+	FloatModelView(new FloatModel(0, 0, 0, 0, nullptr, name, true), this),
+	m_wholeDisplay(numWhole, parent, name, false),
+	m_fractionDisplay(numFrac, parent, name, true),
+	m_mouseMoving(false),
+	m_intStep(false),
+	m_origMousePos(),
+	m_displayOffset(0)
+{
+	layoutSetup();
+}
+
+
+LcdFloatSpinBox::LcdFloatSpinBox(int numWhole, int numFrac, const QString& style, QWidget* parent, const QString& name) :
+	FloatModelView(new FloatModel(0, 0, 0, 0, nullptr, name, true), this),
+	m_wholeDisplay(numWhole, style, parent, name, false),
+	m_fractionDisplay(numFrac, style, parent, name, true),
+	m_mouseMoving(false),
+	m_intStep(false),
+	m_origMousePos(),
+	m_displayOffset(0)
+{
+	layoutSetup(style);
+}
+
+
+void LcdFloatSpinBox::layoutSetup(const QString &style)
+{
+	// Assemble the LCD parts
+	QHBoxLayout *lcdLayout = new QHBoxLayout();
+
+	m_wholeDisplay.setSeamless(false, true);
+	m_fractionDisplay.setSeamless(true, false);
+
+	lcdLayout->addWidget(&m_wholeDisplay);  
+
+	QLabel *dotLabel = new QLabel("", this);
+	QPixmap *dotPixmap = new QPixmap(embed::getIconPixmap(QString("lcd_" + style + "_dot").toUtf8().constData()));
+	dotLabel->setPixmap(dotPixmap->copy(0, 0, dotPixmap->size().width(), dotPixmap->size().height() / 2));
+	lcdLayout->addWidget(dotLabel);
+
+	lcdLayout->addWidget(&m_fractionDisplay);
+
+	lcdLayout->setContentsMargins(0, 0, 0, 0);
+	lcdLayout->setSpacing(0);
+
+	// Add space for label
+	QVBoxLayout *outerLayout = new QVBoxLayout();
+	outerLayout->addLayout(lcdLayout);
+	outerLayout->addSpacing(9);
+	outerLayout->setContentsMargins(0, 0, 0, 0);
+	outerLayout->setSizeConstraint(QLayout::SetFixedSize);
+	this->setLayout(outerLayout);
+}
+
+
+void LcdFloatSpinBox::update()
+{
+	const int whole = static_cast<int>(model()->value());
+	const float fraction = model()->value() - static_cast<int>(model()->value());
```

Comment:
```suggestion
	const float fraction = model()->value() - whole;
```

Compiler might optimize it the way it's written but if not this saves a cast and a method call.